parley/src/layout/line/greedy.rs

```diff
@@ -602,16 +605,50 @@ impl<'a, B: Brush> BreakLines<'a, B> {
                 }
             }
 
+            if quantize {
+                // We round line height to full pixels and then keep track of the overflow.
+                let rounded_line_height = line.metrics.line_height.round();
+                let overflow = line.metrics.line_height - rounded_line_height;
+                line.metrics.line_height = rounded_line_height;
+                // The overflow tracking logic is based on observing Chrome's behavior.
+                // Fractional line height will lead to fluctuating integral perceived line height.
+                if y_overflow <= -0.5 {
+                    // This line will overflow to the previous line.
+                    y -= 1.;
+                    y_overflow += 1.;
+                } else if y_overflow >= 0.5 {
+                    // There will be a 1px gap before the start of the line.
+                    y += 1.;
+                    y_overflow -= 1.;
+                }
+                y_overflow += overflow;
+                // We also mimic Chrome in rounding ascent/descent.
+                line.metrics.ascent = line.metrics.ascent.round();
+                line.metrics.descent = line.metrics.descent.round();
+            }
             line.metrics.leading =
                 line.metrics.line_height - (line.metrics.ascent + line.metrics.descent);
-
+            let leading_below = if quantize {
+                // We mimic Chrome in giving 'above' the remainder in cases of odd leading.
+                // TODO: Test Chrome with a font where descent > ascent to see if it's dynamic.
+                (line.metrics.leading * 0.5).trunc()
+            } else {
+                line.metrics.leading * 0.5
+            };
+            let leading_above = line.metrics.leading - leading_below;
+            // Small line heights will cause leading to be negative.
+            // Negative leadings are correct for baseline calculation, but not for min/max coords.
+            // We clamp leading to zero for the purposes of min/max coords,
+            // which in turn clamps the selection box minimum height to ascent + descent.
+            let negative_leading_above = leading_above.min(0.);
+            let negative_leading_below = leading_below.min(0.);
             // Compute
-            let above = line.metrics.ascent + line.metrics.leading * 0.5;
-            let below = line.metrics.descent + line.metrics.leading * 0.5;
-            line.metrics.min_coord = y;
+            let above = line.metrics.ascent + leading_above;
+            let below = line.metrics.descent + leading_below;
+            line.metrics.min_coord = y + negative_leading_above;
             line.metrics.baseline = y + above;
             y = line.metrics.baseline + below;
-            line.metrics.max_coord = y;
+            line.metrics.max_coord = y - negative_leading_below;
```

Comment:
Previously, we had the (implicit) invariant that summing the `max_coord - min_coord` of every line would, modulo floating-point error, give you the total height of the text. That appears to not be the case here.

The clamping behavior seems to affect what counts as the "logical bounds" of the text, which could affect layout in the various frameworks that use Parley. For instance, decreasing the line height past a certain point will have no effect on the logical height of the text.

Reply:
With this change you have a higher chance of getting the total height of the text. I added this precisely because how incorrectly tiny `max_coord - min_coord` got with small line heights.

Which is to say, what you state is correct - *decreasing the line height past a certain point will have no effect on the logical height of the text* - if we don't counteract the negative leading, then decreasing the line height will incorrectly have an effect on the coords.

Reply:
Is there any way, after this change, of getting the "stylistic line height" of a given line? Right now, you can just do `max_coord - min_coord`, but under this PR, that might give you a number larger than the actual line height.

I need this for egui--I'll need to split up text into multiple layouts, and position each layout after one another vertically by their line heights.

Reply:
So there's:
* line height (which is accessible via `line_height`) - can be quite small (e.g. 0.8em)
* selection box height (`max_coord - min_coord`)

In order to position multiple layouts you would use `line_height`, that is the sum of all the lines.

Now there is the interesting question on whether to support rounding error carry-over in the multiple layout case. Which is to say, you give the error remainder from the previous layout to the next layout as a starting point.

I'll look into this a bit to see how easily we could support such precision in the quantized mode. (*In the raw fractional mode you would do all of that bookkeeping yourself of course.*)

Reply:
Right, but IIRC `line_height` is a style property. Should I just manually accumulate the maximum line height from every styled span within the given line? How will this work once vertical alignment is implemented?

Reply:
> Should I just manually accumulate the maximum line height from every styled span within the given line?

We already do this so you can just access the line's `line_height` which is the maximum of any item on that line.

> How will this work once vertical alignment is implemented?

Hard to say for sure because the implementation doesn't exist, but I like to think that we can keep the line's `line_height` calculation working even with vertical alignment. That might even be a prerequisite to properly implement it.

Reply:
With the latest changes I made, the answer to this question is to use the `Layout::height` method. It returns the sum of all the lines' `line_height`. If `quantize` is `false` then this will give you precisely what you need in vertically stacking multiple `Layout`.

However, if you're depending on the built-in quantization with `quantize` set to `true`, then this approach won't work due to the expectation of `y == 0` being a pixel boundary. This PR has grown rather large already, so while I think it's worth adding as a feature, it's an easy line to draw for splitting the work into a future PR.